README.md
NOTE: Running a SQL Server Inventory will also perform a Windows Inventory!

Requirements
-------------
Before you can document your Windows environment with SQL Power Doc you'll to meet the following requirements:

	Permissions
		- The account used to run SQL Power Doc will need Administrator rights to the OS in order to collect information about the Operating System.

	Windows Machine To Perform Inventory
		- This can be virtual or physical - either way it's recommended that its located on the same physical network as the servers you are collecting information from. SQL Power Doc collects a lot of information and you don't want network latency to become a bottleneck
			
		- PowerShell likes memory; For documenting a 10-20 server environment you'll want at least 2 GB of RAM available. Logical CPU count isn't as important, but SQL Power Doc can split its workload across multiple CPUs when doing it's thing so the more CPUs you've got the faster the job will get done. 

		- The following software needs to be installed: 
			-- Windows PowerShell 2.0 or higher:
				Windows PowerShell 2.0 is available on on all Windows Operating Systems going back to Windows Server 2003 and Windows XP. Chances are you've already got it installed and enabled, but in case you're not sure head over to  http://support.microsoft.com/kb/968929 for a list of requirements and instructions on how to get PowerShell working on your system.

		- Sometimes firewall and group policy restrictions will prevent SQL Power Doc from gathering information from servers. If that's your environment then you'll want to make sure to open up communications for both SQL Server and WMI (Windows Management Instrumentation). Start at  http://msdn.microsoft.com/en-us/library/windows/desktop/aa822854(v=vs.85).aspx for instructions on how to do so.

	Windows Machine To Create The Documentation
		- Usually this will be your laptop or desktop 
		
		- You'll need the following software installed: 
			-- Windows PowerShell 2.0 or higher 
			-- Microsoft Excel 2007 or higher
		
You can use the same machine to collect an inventory and create the documentation as long as it meets all the requirements outlined above.



Configure Windows PowerShell
----------------------------
You'll want to make sure that PowerShell is configured properly on both the machine that you're using to perform the inventory and the machine that's building the documentation. (Repeat: Do this on both machines!)

Set Execution Policy
	By default PowerShell tries to keep you from shooting yourself in the foot by not letting you run scripts that you download from the internet. In PowerShell lingo this is referred to as the Execution Policy (see  http://technet.microsoft.com/en-us/library/hh847748.aspx) and in order to use SQL Power Doc you'll need to change it by following these steps:
		1.Open a PowerShell console in elevated mode:
			Start -> All Programs -> Accessories -> Windows PowerShell -> Windows PowerShell (right click, choose "Run as Administrator")
		
		2.Set the execution policy to allow for remotely signed scripts
			Set-ExecutionPolicy RemoteSigned -Force
		
		3.Exit the PowerShell console

Configure Windows PowerShell Directory
	Now you'll need to create a directory to hold PowerShell code.
		1.Open a new PowerShell console (but not in elevated mode as when you set the execution policy):
			Start -> All Programs -> Accessories -> Windows PowerShell -> Windows PowerShell 

		2.Create PowerShell and PowerShell modules directory in your "My Documents" folder
			New-Item -type directory -path "$([Environment]::GetFolderPath([Environment+SpecialFolder]::MyDocuments))\WindowsPowerShell\Modules"

		3.Exit the PowerShell console

		
Download And Install
--------------------
Grab the latest version of the code from the  downloads page but don't extract the ZIP file yet! Because the file came from the internet it needs to be unblocked or PowerShell gets cranky because it's considered untrusted.

To unblock a file, navigate to it in Windows Explorer, right click, and choose the Properties menu option. On the General tab, click the Unblock button, then click the OK button to close the Properties dialog.

Once the file is unblocked you can extract the contents to the WindowsPowerShell folder (in your "My Documents" directory) that you created in the last step.

Note: Make sure to keep the folder names in the zip file intact so that everything in the Modules folder is extracted into WindowsPowerShell\Modules and the .ps1 files are extracted into the WindowsPowerShell folder.

The Windows Inventory portion of SQL Power Doc will attempt to use the RDS-Manager PowerShell module provided by the Microsoft Remote Desktop Services team. You can download this optional module from  http://gallery.technet.microsoft.com/ScriptCenter/e8c3af96-db10-45b0-88e3-328f087a8700/ . Make sure to save it in the WindowsPowerShell\Modules\RDS-Manager folder (in your "My Documents" directory) that you recently created. It's not the end of the world if it's missing but you'll get more details about users' desktop sessions when it's installed.


Collect Windows Inventory
-------------------------------------------
So far, so good...now it's time to discover your Windows Servers and perform an inventory! In this step you're going to run a PowerShell script which will discover Windows machines on your network (or verify they're running), collect information about them and their underlying OS, and write the results as an XML file that you'll use in the next step.

Start by opening a PowerShell console on the machine that will be collecting the information from your SQL Servers and set your current location to the WindowsPowerShell folder:
	Set-Location "$([Environment]::GetFolderPath([Environment+SpecialFolder]::MyDocuments))\WindowsPowerShell"

Running The Script, Choosing The Right Parameters

	You're going to execute the script .\Get-WindowsInventoryToClixml.ps1 to do all the work but it requires a few parameters to know what to do.

	Discover & Verify Windows Machines

		The first set of parameters define how to find & verify Windows machines.

			Find Windows machines by querying Active Directory DNS for hosts

				-DnsServer
				Valid values are "automatic" or a comma delimited list of AD DNS server IP addresses to query for DNS A records that may be Windows machines.

				-DnsDomain
				Optional. Valid values are "automatic" (the default if this parameter is not specified) or the AD domain name to query for DNS records from.

				-ExcludeSubnet
				Optional. This is a comma delimited list of  CIDR notation subnets to exclude when looking for Windows machines.

				-LimitSubnet
				Optional. This is an inclusive comma delimited list of CIDR notation subnets to limit the scope when looking for Windows machines.

				-ExcludeComputerName
				Optional. This is a comma delimited list of computer names to exclude when looking for Windows machines.

				-PrivateOnly
				Optional. This switch limits the scope to private class A, B, and C IP addresses when looking for Windows machines.


			Find Windows machines by scanning a subnet of IP Addresses

				-Subnet
				Valid values are "automatic" or a comma delimited list of CIDR notation subnets to scan for IPv4 hosts that may be Windows machines.

				-LimitSubnet
				Optional. This is an inclusive comma delimited list of CIDR notation subnets to limit the scope when looking for Windows machines.

				-ExcludeComputerName
				Optional. This is a comma delimited list of computer names to exclude when looking for Windows machines.

				-PrivateOnly
				Optional. This switch limits the scope to private class A, B, and C IP addresses when looking for Windows machines.


			Find Windows machines by computer name

				-ComputerName
				This is a comma delimited list of computer names that may be Windows machines.

				-PrivateOnly
				Optional. This switch limits the scope to private class A, B, and C IP addresses when looking for Windows machines.


	Additional Information To Collect

		By default, SQL Power Doc collects a minimal set of information about each Windows machine it finds. You can collect more information with the following parameter:

			-AdditionalData
			This is a comma delimited list of one or more of the following additional data points to collect:
				AdditionalHardware 
				BIOS 
				DesktopSessions 
				EventLog 
				FullyQualifiedDomainName 
				InstalledApplications 
				InstalledPatches 
				IPRoutes 
				LastLoggedOnUser 
				LocalGroups 
				LocalUserAccounts 
				PowerPlans 
				Printers 
				PrintSpoolerLocation 
				Processes 
				ProductKeys 
				RegistrySize 
				Services 
				Shares 
				StartupCommands 
				WindowsComponents

			Alternatively, you can specify the value "All" to include all of the data points listed above. 

			If this parameter is not provided the default is that none of the listed data points will be included.


	Logging, Output, & Resource Utilization

		Finally, the following parameters control logging, output, and system resources SQL Power Doc will use when finding and collecting information from Windows machines:

			-DirectoryPath
			Optional. A fully qualified directory path where all output will be written. The default value is your "My Documents" folder.

			-LoggingPreference
			Optional. Specifies how much information will be written to a log file (in the same directory as the output file). Valid values are None, Standard, Verbose, and Debug. The default value is None (i.e. no logging).

			-Zip
			Optional. If provided, create a Zip file containing all output in the directory specified by the DirectoryPath parameter.

			-MaxConcurrencyThrottle
			Optional. A number between 1-100 which indicates how many tasks to perform concurrently. The default is the number of logical CPUs present on the OS.

		
	Examples

		The following examples demonstrate how to combine all the parameters together when running the script.

		Example 1:

		.\Get-WindowsInventoryToClixml.psm1 -DNSServer automatic -DNSDomain automatic -PrivateOnly

			Collect an inventory by querying Active Directory for a list of hosts to scan for Windows machines. The list of hosts will be restricted to private IP addresses only.

			The Inventory file will be written to your "My Documents" folder.

			No Log file will be written.


		Example 2:

		.\Get-WindowsInventoryToClixml.psm1 -Subnet 172.20.40.0/28 -LoggingPreference Standard

			Collect an inventory by scanning all hosts in the subnet 172.20.40.0/28 for Windows machines.

			The Inventory and Log files will be written to your "My Documents" folder.

			Standard logging will be used.

		 
		Example 3:

		.\Get-WindowsInventoryToClixml.psm1 -Computername Server1,Server2,Server3

			Collect an inventory by scanning Server1, Server2, and Server3 for Windows machines. 

			The Inventory file will be written to your "My Documents" folder.

			No Log file will be written.
		 

		Example 4:

		.\Get-WindowsInventoryToClixml.psm1 -Computername $env:COMPUTERNAME -AdditionalData None -LoggingPreference Verbose

			Collect an inventory by scanning the local machine for Windows machines.

			Do not collect any data beyond the core set of information.

			The Inventory and Log files will be written to your "My Documents" folder.

			Verbose logging will be used.

 

How Long Will It Take?
----------------------

changelog.txt
PowerShell Inventory Scripts - Kendal Van Dyke.2012.09.28
- Added SQL Agent Job schedules to function Get-SqlAgentJobInformation in module SqlServerDatabaseServerInformation
- Added SQL Agent Job schedules tab to function Export-SqlWindowsInventoryToExcel in module SqlServerInventory


PowerShell Inventory Scripts - Kendal Van Dyke.2012.11.28
- Added formatting to Excel output (now there's options for using Office color themes and dates & numbers are formatted for sorting - no more formula errors!)
	- Recommended color themes:
		- "Waveform" (Office 2010)
		- "Blue" (Office 2013)
- Fixed some issues with cleaning up objects when a Runspace was disposed. This was throwing exceptions in some of my environments. No more!
- Added more stuff I'm collecting from SQL Server and showing in Excel
	- Server Audits
	- Server Audit Specifications
	- Credentials
	- Endpoints
	- Server Triggers
	- VLF counts for transaction log files
	- Server Install & Start Dates
	- Agent Job Steps
	- Agent Job Alerts
	- Agent Job Notifications
	- Agent Alerts
	- Agent Notifications
	- Probably more stuff that I can't remember!
- Updated progress indicators throughout so you have a better idea of what the script's doing
	- (I still have to add better progress reporting to the Windows Inventory piece of this)
- Renamed modules and function calls behind the scenes to refer to the DB Engine as...the DB Engine! (I was calling it Database Server)
- Changed the names of files that are written to disk


PowerShell Inventory Scripts - Kendal Van Dyke.2012.12.19
- Reworked LogHelper module to use synchronized queues for thread safety
- Updated scripts and modules using the LogHelper module to take advantage of synchronized queues for logging
- Fixed SQL Server service port detection in NetworkScan module (it never worked correctly!)
- Updated SQLServerDatabaseEngineInformation module to connect on non-default ports
- Updated SQLServerDatabaseEngineInformation module to collect global trace flags
- Updated SQLServerInventory module to show global trace flags in Excel output


PowerShell Inventory Scripts - Kendal Van Dyke.2012.12.20
- Fixed a problem created by the previous version's addition of port detection that prevented showing Windows information for Database Engine services in Excel


PowerShell Inventory Scripts - Kendal Van Dyke.2012.12.27
- Updated SQLServerDatabaseEngineInformation module to collect Database Mail settings (2005 and higher)
- Updated SQLServerInventory module to show Database Mail Settings in Excel output
- Added date & number formatting to Excel output for Windows Inventory (no more formula errors!)
- Reduced how long it takes to apply formatting to Excel output
- Added error message and line number details to logging in SqlServerDatabaseEngineInformation and WindowsMachineInformation modules
- Refactored Logging:
	- Changed Logging Preferences from 'none','minimal','normal', and 'verbose' to 'none','standard','verbose', and 'debug'
	- Changed Logging Message levels from 'minimal','normal','verbose', and 'debug' to 'information','verbose','debug','error', and 'warning'
	- Added fractions of seconds to log file output
	- Added symbols to Log file to represent message levels
		- ? for Information
		- $ for Verbose
		- * for Debug
		- ! for Error
		- + for Warning
- Updated all modules and scripts to use refactored logging


PowerShell Inventory Scripts - Kendal Van Dyke.2013.01.08
- Removed single tab w\ vertical formatting for Server, Database, and Agent config; replaced with multiple tabs and horizontal formatting (for filtering)
- Updated SQLServerDatabaseEngineInformation module to:
	- Collect Asymmetric and Symmetric key settings (2005 and higher)
	- Collect Resource Governor settings (2005 and higher)
	- Use PSObject instead of hashtable for $ServerInformation
	- Query SERVERPROPERTY for some properties instead of using SMO (to get around inconsistencies between versions of SMO and SQL Server)
	- Add MinimumValue, MaximumValue, IsAdvanced, IsDynamic, FriendlyName, and Description properties to Server Configuration objects coming from sys.configurations
	- Collect Windows Failover Cluster and AlwaysOn configuration information (at the server level)
	- Show Windows Failover Cluster and AlwaysOn configuration information (at the server level)
	- Execute custom T-SQL for enumerating Server Role and Database Role "Member Of" status (previously relied on method calls limited to specific versions of SMO)
- Updated SQLServerInventory module to:
	- Show Asymmetric and Symmetric key settings settings in Excel output
	- Show Windows Failover Cluster and AlwaysOn configuration information in Excel output
	- Show Resource Governor settings in Excel output
	- Limit SQL Agent Job Step command column to 32,767 characters (Excel cell max)
 	- Fixed minor formatting issue with Excel output that wrote blank lines for servers\databases with no information


PowerShell Inventory Scripts - Kendal Van Dyke.2013.01.09
- Fixed an issue in Export-SqlWindowsInventoryToExcel (SqlServerInventory.psm1 ) where lengthy SQL Agent Job Steps would cause an (E_OUTOFMEMORY) exception when applying formatting


PowerShell Inventory Scripts - Kendal Van Dyke.2013.03.08
- General updates
	- Scripts and public functions in all modules now have true PowerShell documentation so they work nicely with Get-Help
	- Lost of minor 1 line fixes I know I made but was too lazy to document at the time and have since forgotten exactly what they were....but they're in there, trust me!
	- Removed -WindowsAuthentication parameter from all scripts and modules. If using SQL authentication provide a username\password, otherwise everything is Windows Authentication by default

Get-SqlServerInventoryToClixml.ps1
<#
	.SYNOPSIS
		Collects comprehensive information about SQL Server instances and their underlying Windows Operating System and saves the result to a file.

	.DESCRIPTION
		This script leverages the NetworkScan, SqlServerInventory, and WindowsInventory modules to scan for and collect comprehensive information about SQL Server instances and their underlying Windows Operating System and save the results to a file in PowerShell's CLIXML format.
		
		This script can find, verify, and collect information by Computer Name, Subnet Scan, or Active Directory DNS query.
		
		This script collects information from SQL Server 2000 or higher running on Windows 2000 or higher and Windows Azure SQL Database (if using SMO 2008 or higher).
		
		This SqlServerInventory and WindowsInventory modules use SQL Server Shared Management Objects (SMO) and Windows Management Instrumentation (WMI) to collect information. As such this script works best when using a version of SMO that matches or is higher than the highest version of each SQL Server instance information is being collected from.
		
		The latest version of SMO can be downloaded from http://www.microsoft.com/en-us/download/details.aspx?id=29065
		Note that SMO also requires the Microsoft SQL Server System CLR Types which can be downloaded from the same page
				
	.PARAMETER  DnsServer
		'Automatic', or the Name or IP address of an Active Directory DNS server to query for a list of hosts to inventory (if there is an instance of SQL Server installed).
		
		When 'Automatic' is specified the function will use WMI queries to discover the current computer's DNS server(s) to query.

	.PARAMETER  DnsDomain
		'Automatic' or the Active Directory domain name to use when querying DNS for a list of hosts.
		
		When 'Automatic' is specified the function will use the current computer's AD domain.
		
		'Automatic' will be used by default if DnsServer is specified but DnsDomain is not provided.
		
	.PARAMETER  Subnet
		'Automatic' or a comma delimited list of subnets (in CIDR notation) to scan for SQL Server instances.
		
		When 'Automatic' is specified the function will use the current computer's IP configuration to determine subnets to scan. 
		
		A quick refresher on CIDR notation:

			BITS	SUBNET MASK			USABLE HOSTS PER SUBNET
			----	---------------		-----------------------
			/20		255.255.240.0		4094
			/21		255.255.248.0		2046 
			/22		255.255.252.0		1022
			/23		255.255.254.0		510 
			/24		255.255.255.0		254 
			/25		255.255.255.128		126 
			/26		255.255.255.192		62
			/27		255.255.255.224		30
			/28		255.255.255.240		14
			/29		255.255.255.248		6
			/30		255.255.255.252		2
			/32		255.255.255.255		1		

	.PARAMETER  ComputerName
		A comma delimited list of computer names to inventory.
	
	.PARAMETER  ExcludeSubnet
		A comma delimited list of subnets (in CIDR notation) to exclude when testing for connectivity.
		
	.PARAMETER  LimitSubnet
		A comma delimited list of subnets (in CIDR notation) to limit the scope of connectivity tests. Only hosts with IP Addresses that fall within the specified subnet(s) will be included in the results.

	.PARAMETER  ExcludeComputerName
		A comma delimited list of computer names to exclude when testing for connectivity. Wildcards are accepted.
		
		An attempt will be made to resolve the IP Address(es) for each computer in this list and those addresses will also be used when determining if a host should be included or excluded when testing for connectivity.		

	.PARAMETER  Username
		SQL Server username to use when connecting to instances. 
		
		Windows authentication will be used to connect if this parameter is not provided.

	.PARAMETER  Password
		SQL Server password to use when connecting to instances.

	.PARAMETER  MaxConcurrencyThrottle
		Number between 1-100 to indicate how many instances to collect information from concurrently.

		If not provided then the number of logical CPUs present to your session will be used.

	.PARAMETER  PrivateOnly
		Restrict inventory to instances on private class A, B, or C IP addresses

	.PARAMETER  DirectoryPath
		Specifies the literal path to the directory where the inventory file and log file will be written.
		
		If not specified then the script defaults to your "My Documents" folder.		

	.PARAMETER  LoggingPreference
		Specifies the logging verbosity to use when writing log entries.
		
		Valid values include: None, Standard, Verbose, and Debug.
		
		The default value is "None"

	.PARAMETER  Zip
		Combine the Inventory and Log files into a single compressed ZIP file. This is useful for transferring the output of an inventory to another machine for further analysis.
		
	.PARAMETER  IncludeDatabaseObjectPermissions
		Includes database object level permissions (System object permissions included only if -IncludeDatabaseSystemObjects is also provided)

	.PARAMETER  IncludeDatabaseObjectInformation
		Includes database object information (but does not include system objects)

	.PARAMETER  IncludeDatabaseSystemObjects
		Include system objects when retrieving database object information. 
		
		This has no effect if neither -IncludeDatabaseObjectInformation nor -IncludeDatabaseObjectPermissions are specified.

	.EXAMPLE
		.\Get-SqlServerInventoryToClixml.ps1 -DNSServer automatic -DNSDomain automatic -PrivateOnly
		
		Description
		-----------
		Collect an inventory by querying Active Directory for a list of hosts to scan for SQL Server instances. The list of hosts will be restricted to private IP addresses only.
		
		Windows Authentication will be used to connect to each instance.
		
		Database objects will NOT be included in the results.
		
		The Inventory file will be written to your "My Documents" folder.
		
		No log file will be written.

	.EXAMPLE
		.\Get-SqlServerInventoryToClixml.ps1 -Subnet 172.20.40.0/28 -Username sa -Password BetterNotBeBlank
		
		Description
		-----------
		Collect an inventory by scanning all hosts in the subnet 172.20.40.0/28 for SQL Server instances.
		
		SQL authentication (username = "sa", password = "BetterNotBeBlank") will be used to connect to the instance.
		
		Database objects will NOT be included in the results.

		The Inventory file will be written to your "My Documents" folder.
		
		No log file will be written.		
		
	.EXAMPLE
		.\Get-SqlServerInventoryToClixml.ps1 -Computername Server1,Server2,Server3 -LoggingPreference Standard
		
		Description
		-----------
		Collect an inventory by scanning Server1, Server2, and Server3 for SQL Server instances.
		
		Windows Authentication will be used to connect to the instance.
		
		Database objects will NOT be included in the results.

		The Inventory file will be written to your "My Documents" folder.

		Standard logging will be used.		

	.EXAMPLE
		.\Get-SqlServerInventoryToClixml.ps1 -Computername $env:COMPUTERNAME -IncludeDatabaseObjectInformation -LoggingPreference Verbose
		
		Description
		-----------
		Collect an inventory by scanning the local machine for SQL Server instances.
		
		Windows Authentication will be used to connect to the instance.
		
		Database objects (EXCLUDING system objects) will be included in the results.

		The Inventory file will be written to your "My Documents" folder.

		Verbose logging will be used.			
		
	.EXAMPLE
		.\Get-SqlServerInventoryToClixml.ps1 -Computername $env:COMPUTERNAME -IncludeDatabaseObjectInformation -IncludeDatabaseSystemObjects

		Description
		-----------
		Collect an inventory by scanning the local machine for SQL Server instances.
		
		Windows Authentication will be used to connect to the instance.
		
		Database objects (INCLUDING system objects) will be included in the results.
		

	.OUTPUTS
		System.Management.Automation.PSObject

	.NOTES

	.LINK
		.\Convert-SqlServerInventoryClixmlToExcel.ps1

#>
[cmdletBinding(SupportsShouldProcess=$True, DefaultParametersetName='computername_WindowsAuthentication')]
param(
	[Parameter(Mandatory=$true, ParameterSetName='dns_SQLAuthentication', HelpMessage='DNS Server(s)')]
	[Parameter(Mandatory=$true, ParameterSetName='dns_WindowsAuthentication', HelpMessage='DNS Server(s)')]
	[alias('dns')]
	[ValidatePattern('^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$|^auto$|^automatic$')]
	[string[]]
	$DnsServer = 'automatic'
	,
	[Parameter(Mandatory=$false, ParameterSetName='dns_SQLAuthentication', HelpMessage='DNS Domain Name')] 
	[Parameter(Mandatory=$false, ParameterSetName='dns_WindowsAuthentication', HelpMessage='DNS Domain Name')] 
	[alias('domain')]
	[string]
	$DnsDomain = 'automatic'
	,
	[Parameter(Mandatory=$true, ParameterSetName='subnet_SQLAuthentication', HelpMessage='Subnet (in CIDR notation)')] 
	[Parameter(Mandatory=$true, ParameterSetName='subnet_WindowsAuthentication', HelpMessage='Subnet (in CIDR notation)')] 
	[ValidatePattern('^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)[\\/]\d{1,2}$|^auto$|^automatic$')]
	[string[]]
	$Subnet = 'automatic'
	,
	[Parameter(Mandatory=$true, ParameterSetName='computername_SQLAuthentication', HelpMessage='Computer Name(s)')] 
	[Parameter(Mandatory=$true, ParameterSetName='computername_WindowsAuthentication', HelpMessage='Computer Name(s)')] 
	[alias('computer')]
	[string[]]
	$ComputerName
	,
	[Parameter(Mandatory=$false, ParameterSetName='dns_SQLAuthentication')]
	[Parameter(Mandatory=$false, ParameterSetName='dns_WindowsAuthentication')]
	[Parameter(Mandatory=$false, ParameterSetName='subnet_SQLAuthentication')]
	[Parameter(Mandatory=$false, ParameterSetName='subnet_WindowsAuthentication')]
	[ValidatePattern('^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)[\\/]\d{1,2}$')]
	[string[]]
	$ExcludeSubnet
	,
	[Parameter(Mandatory=$false, ParameterSetName='dns_SQLAuthentication')]
	[Parameter(Mandatory=$false, ParameterSetName='dns_WindowsAuthentication')]
	[ValidatePattern('^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)[\\/]\d{1,2}$')]
	[string[]]
	$LimitSubnet
	,
	[Parameter(Mandatory=$false, ParameterSetName='dns_SQLAuthentication')]
	[Parameter(Mandatory=$false, ParameterSetName='dns_WindowsAuthentication')]
	[Parameter(Mandatory=$false, ParameterSetName='subnet_SQLAuthentication')]
	[Parameter(Mandatory=$false, ParameterSetName='subnet_WindowsAuthentication')]
	[string[]]
	$ExcludeComputerName
	, 
	[Parameter(Mandatory=$true, ParameterSetName='dns_SQLAuthentication') ]
	[Parameter(Mandatory=$true, ParameterSetName='subnet_SQLAuthentication') ]
	[Parameter(Mandatory=$true, ParameterSetName='computername_SQLAuthentication') ]
	[ValidateNotNull()]
	[System.String]
	$Username
	,
	[Parameter(Mandatory=$true, ParameterSetName='dns_SQLAuthentication') ]
	[Parameter(Mandatory=$true, ParameterSetName='subnet_SQLAuthentication') ]
	[Parameter(Mandatory=$true, ParameterSetName='computername_SQLAuthentication') ]
	[ValidateNotNull()]
	[System.String]
	$Password
	,
	[Parameter(Mandatory=$false)] 
	[ValidateRange(1,100)]
	[byte]
	$MaxConcurrencyThrottle = $env:NUMBER_OF_PROCESSORS
	,
	[Parameter(Mandatory=$false)] 
	[switch]
	$PrivateOnly = $false
	,
	[Parameter(Mandatory=$false)] 
	[alias('Directory','Path')]
	[ValidateNotNullOrEmpty()]
	[string]
	$DirectoryPath = ([Environment]::GetFolderPath([Environment+SpecialFolder]::MyDocuments))
	,
	[Parameter(Mandatory=$false)] 
	[alias('LogLevel')]
	[ValidateSet('none','standard','verbose','debug')]
	[string]
	$LoggingPreference = 'none'
	,
	[Parameter(Mandatory=$false)] 
	[switch]
	$Zip = $false
	,
	[Parameter(Mandatory=$false)]
	[switch]
	$IncludeDatabaseObjectPermissions = $false
	,
	[Parameter(Mandatory=$false)]
	[alias('IncludeDbObjects','IncludeDatabaseObjects')]
	[switch]
	$IncludeDatabaseObjectInformation = $false
	,
	[Parameter(Mandatory=$false)]
	[switch]
	$IncludeDatabaseSystemObjects = $false
) 


######################
# FUNCTIONS
######################

function Write-LogMessage {
	[CmdletBinding()]
	param(
		[Parameter(Position=0, Mandatory=$true)]
		[ValidateNotNullOrEmpty()]
		[System.String]
		$Message
		,
		[Parameter(Position=1, Mandatory=$true)] 
		[alias('level')]
		[ValidateSet('information','verbose','debug','error','warning')]
		[System.String]
		$MessageLevel
	)
	try {
		if ((Test-Path -Path 'function:Write-Log') -eq $true) {
			Write-Log -Message $Message -MessageLevel $MessageLevel
		} else {
			Write-Host $Message
		}
	}
	catch {
		Throw
	}
}


function Test-FileIsOpen {
	[CmdletBinding()]
	[OutputType([System.Boolean])]
	param(
		[Parameter(Position=0, Mandatory=$true)]
		[ValidateNotNullOrEmpty()]
		[System.String]
		$Path
	)
	process {
		$FileIsOpen = $false
		$Filestream = $null

		try {
			$Filestream = [System.IO.File]::Open($ZipFilePath, [System.IO.FileMode]::Open, [System.IO.FileAccess]::Read, [System.IO.FileShare]::None)
			$Filestream.Close()
			Write-Output $false
		}
		catch {
			Write-Output $true
		}
	}
}


######################
# VARIABLES
######################

$Inventory = $null
$ParameterHash = $null
[String]$ZipFilePath = $null
$ZipFile = $null

$BasePath = (Join-Path -Path $DirectoryPath -ChildPath ('SQL Server Inventory - ' + (Get-Date -Format 'yyyy-MM-dd-HH-mm')))
$CliXmlPath = [System.IO.Path]::ChangeExtension($BasePath, 'xml.gz')
$LogPath = [System.IO.Path]::ChangeExtension($BasePath, 'log')
$ZipFilePath = [System.IO.Path]::ChangeExtension($BasePath,'zip')

# Fallback in case value isn't supplied or somehow missing from the environment variables
if (-not $MaxConcurrencyThrottle) { $MaxConcurrencyThrottle = 1 }

######################
# BEGIN SCRIPT
######################

# Import Modules that we need
Import-Module -Name LogHelper, SqlServerInventory


# Set logging variables
Set-LogFile -Path $LogPath
Set-LoggingPreference -Preference $LoggingPreference

Write-LogMessage -Message "Starting Script: $($MyInvocation.MyCommand.Path)" -MessageLevel Information
<#
$PsCmdlet.MyInvocation.MyCommand.Parameters.Values | Where-Object { $_.ParameterType.Name -ine 'actionpreference' } | ForEach-Object {
    $Param = Invoke-Expression $('[' + $(if ($_.ParameterType.Name -ieq 'SwitchParameter') { 'Boolean' } else { $_.ParameterType.Name }) + ']$' + $_.Name)
	Write-LogMessage -Message "`t-$($_.Name): $Param" -MessageLevel Information
} 
#>

# Build inventory collection command parameters
$ParameterHash = @{
	MaxConcurrencyThrottle = $MaxConcurrencyThrottle
	PrivateOnly = $PrivateOnly
	IncludeDatabaseObjectPermissions = $IncludeDatabaseObjectPermissions
	IncludeDatabaseObjectInformation = $IncludeDatabaseObjectInformation
	IncludeDatabaseSystemObjects = $IncludeDatabaseSystemObjects
}

switch ($PsCmdlet.ParameterSetName) {
	'dns_SQLAuthentication' {
		$ParameterHash.Add('Username',$Username)
		$ParameterHash.Add('Password',$Password)
		$ParameterHash.Add('DnsServer',$DnsServer)
		$ParameterHash.Add('DnsDomain',$DnsDomain)
		if ($ExcludeSubnet) { $ParameterHash.Add('ExcludeSubnet',$ExcludeSubnet) }
		if ($LimitSubnet) { $ParameterHash.Add('IncludeSubnet',$LimitSubnet) }
		if ($ExcludeComputerName) { $ParameterHash.Add('ExcludeComputerName',$ExcludeComputerName) }
	}
	'dns_WindowsAuthentication' {
		$ParameterHash.Add('DnsServer',$DnsServer)
		$ParameterHash.Add('DnsDomain',$DnsDomain)
		if ($ExcludeSubnet) { $ParameterHash.Add('ExcludeSubnet',$ExcludeSubnet) }
		if ($LimitSubnet) { $ParameterHash.Add('IncludeSubnet',$LimitSubnet) }
		if ($ExcludeComputerName) { $ParameterHash.Add('ExcludeComputerName',$ExcludeComputerName) }
	}
	'subnet_SQLAuthentication' {
		$ParameterHash.Add('Username',$Username)
		$ParameterHash.Add('Password',$Password)
		$ParameterHash.Add('Subnet',$Subnet)
		if ($ExcludeSubnet) { $ParameterHash.Add('ExcludeSubnet',$ExcludeSubnet) }
		if ($ExcludeComputerName) { $ParameterHash.Add('ExcludeComputerName',$ExcludeComputerName) }
	}
	'subnet_WindowsAuthentication' {
		$ParameterHash.Add('Subnet',$Subnet)
		if ($ExcludeSubnet) { $ParameterHash.Add('ExcludeSubnet',$ExcludeSubnet) }
		if ($ExcludeComputerName) { $ParameterHash.Add('ExcludeComputerName',$ExcludeComputerName) }
	}
	'computername_SQLAuthentication' {
		$ParameterHash.Add('Username',$Username)
		$ParameterHash.Add('Password',$Password)
		$ParameterHash.Add('ComputerName',$ComputerName)
	}
	'computername_WindowsAuthentication' {
		$ParameterHash.Add('ComputerName',$ComputerName)
	}
}


# Collect inventory and export results to Excel (if there are results in the inventory collection)
$Inventory = Get-SqlServerInventory @ParameterHash

Write-LogMessage -Message 'Writing Inventory to disk' -MessageLevel Information

if ($Inventory.DatabaseServerScanSuccessCount -gt 0) {
	#Compress-SqlServerInventory -SqlServerInventory ([REF]$Inventory)
	#$Inventory | Export-Clixml -Path $CliXmlPath -Force -Depth 100 -Encoding UTF8

	$Inventory | Export-SqlServerInventoryToGzClixml -Path $CliXmlPath

	#Compress-SqlServerInventory -SqlServerInventory ([REF]$Inventory) | Export-Clixml -Path $CliXmlPath -Force -Depth 100 -Encoding UTF8
	#$_ | Export-Clixml -Path $CliXmlPath -Force -Depth 100 -Encoding UTF8
} else {
	Write-LogMessage -Message 'No machines found!' -MessageLevel Warning
}

Write-LogMessage -Message "End Script: $($MyInvocation.MyCommand.Path)" -MessageLevel Information


# Try to create ZIP archive if the option was specified and a log or CliXml file was created
if (
	$Zip -eq $true -and 
	(
		$(Test-Path -Path $LogPath) -or 
		$(Test-Path -Path $CliXmlPath)
	)
) {

	# Create the zip file; if it already exists write a message to the console and skip this part
	if ($(Test-Path -Path $ZipFilePath) -ne $true) {

		Set-Content -Path $ZipFilePath -Value ("PK" + [char]5 + [char]6 + ("$([char]0)" * 18))
		$ZipFile = (New-Object -ComObject Shell.Application).NameSpace($ZipFilePath)

		# Add log file if it exists
		if (($LoggingPreference -ine 'none') -and (Test-Path -Path $LogPath)) { 
			$ZipFile.CopyHere($LogPath)

			Start-Sleep -Milliseconds 500
			while (Test-FileIsOpen -Path $ZipFilePath) {
				Start-Sleep -Seconds 1
			}
		}

		# Add CliXml file if it exists
		if (Test-Path -Path $CliXmlPath) { 
			$ZipFile.CopyHere($CliXmlPath) 

			Start-Sleep -Milliseconds 500
			while (Test-FileIsOpen -Path $ZipFilePath) {
				Start-Sleep -Seconds 1
			} 
		}

		# Remove reference to zip file
		$ZipFile = $null

	} else {
		Write-LogMessage -Message "Unable to compress files - '$ZipFilePath' already exists" -MessageLevel Error
	}
}


# Remove Variables
Remove-Variable -Name Inventory, ParameterHash, ZipFilePath, ZipFile, BasePath, CliXmlPath, LogPath

# Remove Modules
Remove-Module -Name SqlServerInventory, LogHelper

# Call garbage collector
[System.GC]::Collect()
